specs/core/0_beacon-chain.md

```diff
@@ -1772,135 +1773,310 @@ For each `transfer` in `block.body.transfers`:
+    all_attestations = get_current_attestations(state) + get_previous_attestations(state)
+    valid_attestations = [
+        a for a in all_attestations if a.data.latest_crosslink == state.latest_crosslinks[shard]
+    ]
+    all_roots = [a.data.crosslink_data_root for a in valid_attestations]
+    
+    def get_attestations_for(root) -> List[PendingAttestation]:
+        return [a for a in valid_attestations if a.data.crosslink_data_root == root]
+        
+    winning_root = max(all_roots, key=lambda r: get_attesting_balance(get_attestations_for(r)))
+    
+    return winning_root, get_attesting_indices(get_attestations_for(winning_root))
+```
 
-* If `eth1_data_vote.vote_count * 2 > EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH` for some `eth1_data_vote` in `state.eth1_data_votes` (ie. more than half the votes in this voting period were for that value), set `state.latest_eth1_data = eth1_data_vote.eth1_data`.
-* Set `state.eth1_data_votes = []`.
+```python
+def earliest_attestation(state: BeaconState, validator_index: ValidatorIndex) -> PendingAttestation:
+    return min([
+        a for a in state.previous_epoch_attestations if
+        validator_index in get_attestation_participants(state, a.data, a.aggregation_bitfield)
+    ], key=lambda a: a.inclusion_slot)
+```
 
-#### Justification
+```python
+def inclusion_slot(state: BeaconState, validator_index: ValidatorIndex) -> Slot:
+    return earliest_attestation(state, validator_index).inclusion_slot
+```
 
-First, update the justification bitfield:
+```python
+def inclusion_distance(state: BeaconState, validator_index: ValidatorIndex) -> int:
+    attestation = earliest_attestation(state, validator_index)
+    return attestation.inclusion_slot - attestation.data.slot
+```
 
-* Let `new_justified_epoch = state.justified_epoch`.
-* Set `state.justification_bitfield = state.justification_bitfield << 1`.
-* Set `state.justification_bitfield |= 2` and `new_justified_epoch = previous_epoch` if `3 * previous_epoch_boundary_attesting_balance >= 2 * previous_total_balance`.
-* Set `state.justification_bitfield |= 1` and `new_justified_epoch = current_epoch` if `3 * current_epoch_boundary_attesting_balance >= 2 * current_total_balance`.
+#### Justification
 
-Next, update last finalized epoch if possible:
+Run the following function:
+
+```python
+def update_justification_and_finalization(state: BeaconState) -> None:
+
+    new_justified_epoch = state.justified_epoch
+    # Rotate the justification bitfield up one epoch to make room for the current epoch
+    state.justification_bitfield <<= 1
+    # If the previous epoch gets justified, fill the second last bit
+    previous_boundary_attesting_balance = get_attesting_balance(get_previous_epoch_boundary_attestations(state))
+    if previous_boundary_attesting_balance * 3 >= get_previous_total_balance(state) * 2:
+        new_justified_epoch = get_current_epoch(state) - 1
```

Comment:
The `previous_epoch`s, which were defined by `get_previous_epoch`, are all replaced with `current_epoch -1` in this function body. This would be problematic in the genesis epoch.

Reply:
@djrtwo and I are talking about this; it seems likely that setting previous to equal current minus 1 in all cases could potentially be ok.